Port resvg to `harfrust` and `fontations`

nicoburns opened the pull request:
## Changes made

- Replaces `rustybuzz` with `harfrust`
- Replaces `ttf_parser` with `skrifa`

## Notes

- This is failing 23 tests. These probably need to be manually checked. I'm waiting on https://github.com/linebender/resvg/pull/954 to investigate these.
- This requires an MSRV bump from `1.65` to `1.85`
- This will represent a binary size bump for users of `resvg`.

## Production profile

The analysis below uses the following "production" profile:

```toml
[profile.production]
inherits = "release"
opt-level = 3
debug = false
lto = true
codegen-units = 1
strip = true
incremental = false
```

## Binary size analysis:

According to `cargo-bloat` (not the most accurate):

  - `harfrust` and `rustybuzz` are both ~250kb
  - `ttf_parser` and `read-fonts` are both ~100kb
  - However, `ttf_parser` includes the metrics outlining functionality of `skrifa`. And skrifa is another ~250kb.

Compiling the CLI (which is a much more useful test) with the production profile above (`cargo build --profile production --bin resvg`) I get:

- `2.4mb` for `main`
- `3.0mb` for this PR

**So this PR causes a binary size increase of around `600kb`**

## Compile time analysis

Compile times on an M1 Pro (8+2 cores) are shown below (`cargo build`):

| Profile | `main` | This PR |
| --- | ---| --- |
| `debug` | 4.68s | 9.09s |
| `release` | 7.64s | 14.84s |
| `production` | 24.59s | 34.00s |

**So this PR roughly doubles compile times at regular opt settings**
(with the effect being less pronounced with LTO and other production optimisation settings enabled)

nicoburns committed:
Use skrifa instead of ttf_parser

Signed-off-by: Nico Burns <nico@nicoburns.com>

nicoburns committed:
Use harfrust instead of rustybuzz

Signed-off-by: Nico Burns <nico@nicoburns.com>

LaurenzV commented:
Yeah I think we first have to figure out how to deal with fontdb before doing anything with resvg. I don’t think RazrFalcon would want to switch it, and I’m not sure that fontique supports everything we need?

RazrFalcon commented:
It's an interesting experiment, but I don't see a point in it. Is there a reason in switching to `harfrust`? 
Last time I have checked it was still a beta, while `rustybuzz` (with all its faults) is a pretty robust solution.

PS: even if `harfrust` is faster, it doesn't matter either, because shaping isn't a bottleneck, lack of glyphs caching is.

RazrFalcon commented:
Oh wait, it depends on proc-macros/syn?! Burn it with fire! No proc-macros in `resvg`.

xStrom commented on .github/workflows/main.yml:
```diff
@@ -65,7 +65,7 @@ jobs:
         uses: dtolnay/rust-toolchain@stable
         with:
-          toolchain: 1.67.1
+          toolchain: 1.79.0
```
Shouldn't this (and the other `rust-version` modifications) be 1.75.0 not 79?

xStrom commented:
> Last time I have checked it was still a beta

Not sure about 'beta' type labels, but the first release ([0.1.0](https://github.com/harfbuzz/harfrust/releases/tag/0.1.0)) of HarfRust happened on June 10, marking a significant milestone.

> It's an interesting experiment, but I don't see a point in it. Is there a reason in switching to `harfrust`?

For an immediate effect, it seems to be far more compliant when tested against the HarfBuzz test suite.

To quote Chad:
> We’re still mulling over some public API changes and need to do some serious performance work but this is fully backed by fontations, up to date with HarfBuzz 11.2.1, and passes a significantly larger portion of the HB test suite than RustyBuzz: latest CI shows 5911 tests passing for HarfRust vs 2267 for RustyBuzz.

And the [HarfRust introduction post](https://docs.google.com/document/d/1aH_waagdEM5UhslQxCeFEb82ECBhPlZjy5_MwLNLBYo/preview) by Behdad contains the following claim:

> At the time of this writing, HarfRust passes most of the HarfBuzz shaping test suites, failing just 23 tests out of more than 6,000.

---

Moving forward, HarfRust will receive active development and continuous sync with HarfBuzz, while RustyBuzz is likely to only see critical fixes.

RazrFalcon commented:
>latest CI shows 5911 tests passing for HarfRust vs 2267 for RustyBuzz.

Apples and oranges, but sure.

Either way, proc-macros is the current major roadblock.

mattfbacon commented:
No it wouldn't be at all, since the bytemuck derive macros generate unsafe code so they would need to allow unsafe code back into the codebase.

But what is your plan here? Earlier you said wait, but for what?

behdad commented:
> The geometric mean of the last result is ~1.16, meaning Harfrust is now only 16% slower than Harfbuzz.

It's even better than that. See the comment:
https://github.com/harfbuzz/harfrust/discussions/257#discussioncomment-14363176

In short, there's some hb-harfrust overhead included in the spreadsheet number.

nicoburns commented:
Diff for the colv1 test https://www.diffchecker.com/V8j95jlx with some manual fixups (removing groups - although it could also be those causing the differences?). I'm seeing some negations (which possibly cancel each other out?), and also that skrifa is applying some opacities to groups whereas ttf-parser applies them directly to the paths.

LaurenzV commented:
It's possible the difference stems from the fact that the opacities are applied differently. When applying directly to a path, it will simply draw a path with that, while when applying to a group resvg will do layer isolation and apply the opacity that way instead, so you could try if this can be fixed by applying it directly.

nicoburns committed:
Update test snapshots

nicoburns committed:
Better output for failing usvg tests

nicoburns committed:
Update usvg test assertions

nicoburns commented:
Ok, I've managed to fix the CBDT offset issue (I admit more by trial-and-error than by principle), and I've also managed to fix to fix the COLRv0 missing glyph parts (this was a silly bug where string was being thrown away rather than persisted).

That leaves:

### The COLRv1 and Zalgo tests. Diff output from the test runner for these tests is as below:

<img width="900" height="300" alt="tests_text_color-font_colrv1" src="https://github.com/user-attachments/assets/fc96b358-a00d-432f-9fcd-d1c533cf26b4" />

<img width="900" height="300" alt="tests_text_text_zalgo" src="https://github.com/user-attachments/assets/554ebf6b-0ac7-4b0a-8d27-26ac547ea089" />

For the COLRv1 tests, see also the diff in the generated "lowered" SVG output (https://www.diffchecker.com/V8j95jlx)

@dfrg Do you have an opinion on whether these represent a genuine bug, a fixing of a bug, or a valid difference in interpretation?

### The negations

This seems to be negated output for each y-coord in path, which is then cancelled out by a negated y coordinate in a transform matrix. This is probably fine?

### Missing path segments

Some of the text-related `usvg` "resave" tests (parse SVG then rereserialize back to SVG) are failing. Some sample diffs:

- text_simple https://www.diffchecker.com/e2SQWqnm/
- complex_text https://www.diffchecker.com/tQk27XmG/
- preserve_text_simple_case https://www.diffchecker.com/Ip5F1whl/
- text_with_generated_gradients https://www.diffchecker.com/gYS1sYk2
- clip_path_with_text https://www.diffchecker.com/tDIazl6O/

The issue seems to be that the last segment of some (but not all) paths is being dropped in the Skrifa implementation. Although this seems to make no visual difference to the rendered SVG (see github diff), so perhaps that's also fine?

---

RazrFalcon commented:
>As some months have passed another benefit has appeared - performance. 

Good. It would not affect `resvg` in any way, because shaping was never a bottleneck, but still a progress.
And `rustybuzz` is dead anyways.

nicoburns committed:
Use harfrust 0.3.1 and skrifa 0.37

Signed-off-by: Nico Burns <nico@nicoburns.com>

nicoburns committed:
Upgrade to HarfRust 0.5.0, Skrifa 0.40.0

Signed-off-by: Nico Burns <nico@nicoburns.com>

nicoburns committed:
Implement font variations in layout.rs

Signed-off-by: Nico Burns <nico@nicoburns.com>

nicoburns committed:
Enable variations in flatten.rs

Signed-off-by: Nico Burns <nico@nicoburns.com>

nicoburns committed:
Update snapshots to reflect state of PR

Enivex commented:
I don't know what the status of this is, but harfrust 0.6.0 was released today

https://github.com/harfbuzz/harfrust/releases/tag/0.6.0

nicoburns commented:
>  harfrust 0.6.0 was released today

Indeed, although it's just a dep bump release.

> I don't know what the status of this is

The remaining blocker is switching resvg to use `fontique` instead of `fontdb` (see: https://xi.zulipchat.com/#narrow/channel/465085-resvg/topic/Moving.20resvg.20to.20fontations.2Fharfrust/with/572344735). I have some half-baked (not even compiling yet) changes for this locally, but it's unfortunately not top of my priority queue right now.

Enivex commented:
> Indeed, although it's just a dep bump release.

The numbering is a bit strange. There were more changes in 0.5.1 and 0.5.2.

Edit: I just realized that you were the one who bumped it 😂

laurmaedje commented:
> The numbering is a bit strange. There were more changes in 0.5.1 and 0.5.2.

Just normal SemVer I would assume. It looks like `read_fonts` is in the public API of harfrust, so bumping its minor version is a breaking change for harfrust users.

nicoburns committed:
Upgrade to Skirfa 0.42, HarfRust 0.8.4

Signed-off-by: Nico Burns <nico@nicoburns.com>

nicoburns committed:
Update snapshots

Signed-off-by: Nico Burns <nico@nicoburns.com>

nicoburns committed:
Upgrade deps: HarfRust 0.12, Skrifa 0.44

Signed-off-by: Nico Burns <nico@nicoburns.com>

LaurenzV commented on crates/resvg/tests/tests/text/color-font/colrv1.png:
Did you run oxipng on the new images?

nicoburns replied:
I had not, but I have now :)

LaurenzV commented on crates/usvg/Cargo.toml:
```diff
@@ -54,7 +52,7 @@ default = ["svgz", "text", "system-fonts", "memmap-fonts", "writer"]
 svgz = ["flate2"]
 # Enables text-to-path conversion support.
 # Adds around 400KiB to your binary.
```
Is this comment still accurate?

nicoburns replied:
It seems to be more like 1400kb now!

LaurenzV commented on crates/usvg/src/text/transform.rs:
```diff
@@ -0,0 +1,17 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+pub(crate) fn skrifa_to_tsp_transform(t: skrifa::color::Transform) -> tiny_skia_path::Transform {
+    tiny_skia_path::Transform::from_row(t.xx, t.yx, t.xy, t.yy, t.dx, t.dy)
```
Have you double-checked that the yx and xy are in the correct order? I keep mixing them up. 😄 Might be worth trying to switch them to ensure that some test fails. Otherwise, we should double-check.

nicoburns replied:
Not recently, but I did go quite deep on this originally. There's a whole zulip thread somewhere with matchup table for the different transform types of a few different libraries.

LaurenzV commented on crates/usvg/src/text/mod.rs:
```diff
@@ -171,16 +182,10 @@ impl FontResolver<'_> {
                 let base_family = base_face
                     .families
                     .iter()
-                    .find(|f| f.1 == fontdb::Language::English_UnitedStates)
```
Those are the same right?

nicoburns replied:
This change has disappeared now that we're using fontdb with inlined ttf-parser rather than read-fonts.

LaurenzV commented on crates/usvg/src/text/layout.rs:
```diff
@@ -1209,43 +1212,47 @@ impl DatabaseExt for Database {
+            let font = skrifa::FontRef::from_index(data, face_index).ok()?;
 
-            let units_per_em = NonZeroU16::new(font.units_per_em())?;
+            // FIXME: Set size and location
```
Meaning we are currently doing this wrong and don't get the metrics for the right variations?

nicoburns replied:
Yeah. I should probably fill this in. Context: when I started this PR, `main` didn't do variations either.

nicoburns replied:
I have now filled in variable font support (AI generated with Fable)

LaurenzV commented on crates/usvg/src/text/colr.rs:
```diff
@@ -41,71 +49,75 @@ impl ttf_parser::OutlineBuilder for Builder<'_> {
         for stop in stops {
+            let color = self
+                .palette_index_to_color(stop.palette_index, stop.alpha)
+                .unwrap();
```
Maybe defensively fall back to black?

LaurenzV commented on crates/usvg/src/text/colr.rs:
```diff
@@ -184,47 +206,29 @@ impl<'a> GlyphPainter<'a> {
+        _color_stops: &[ColorStop],
+        _extend: Extend,
+    ) {
         println!("Warning: sweep gradients are not supported.");
```
This should probably be a log right?

nicoburns committed:
Use fontdb with ttf-parser inlined

Signed-off-by: Nico Burns <nico@nicoburns.com>

LaurenzV commented on crates/usvg/src/text/colr.rs:
```diff
@@ -245,14 +249,52 @@ impl GlyphPainter<'_> {
+            self.foreground_color
+        } else {
+            let cpal = self.font.cpal().ok()?;
+            let color = cpal.color_records_array()?.ok()?[palette_index as usize];
```
I think it needs to be like this:

```suggestion
let palettes = self.font.color_palettes();
            let color = *palettes.get(0)?.colors().get(palette_index as usize)?;
```

LaurenzV commented on crates/usvg/src/text/colr.rs:
```diff
@@ -245,14 +249,52 @@ impl GlyphPainter<'_> {
+impl<'a> skrifa::color::ColorPainter for GlyphPainter<'a> {
+    fn push_transform(&mut self, transform: Transform) {
+        self.transforms_stack.push(self.transform);
+        self.transform = transform * self.transform;
```
That's the wrong order no?

nicoburns replied:
Probably! I find transform application order very confusing.

nicoburns replied:
I have swapped the multiplication order here (I think that's what you meant?)

LaurenzV commented on crates/usvg/src/text/colr.rs:
```diff
@@ -245,14 +249,52 @@ impl GlyphPainter<'_> {
-        let mut builder = Builder(self.path_buf);
-        match self.face.outline_glyph(glyph_id, &mut builder) {
-            Some(v) => v,
+        let mut builder = Builder(&mut self.path_buf);
```
I think we need to also apply the glyph here? Previously it was split into `outline_glyph` and `push_clip`, but I think for the skrifa one you literally need "push this to the clip stack".

nicoburns replied:
Sorry - I don't think I understand this comment! (what is "apply the glyph"?)

LaurenzV commented on crates/usvg/src/text/colr.rs:
```diff
@@ -280,64 +371,20 @@ impl<'a> ttf_parser::colr::Painter<'a> for GlyphPainter<'a> {
             }
         };
+
+        self.svg.start_element("g");
```
Since we push a group here we need to override the `pop_layer` impl from the trait as well to close it.

nicoburns replied:
Good point. Added.

LaurenzV reviewed (commented):
Not fully done yet but some first comments.

nicoburns committed:
Fix COLR transform composition order

Skrifa's Matrix multiplication chains right-to-left (C = A * B applies B
then A), so nested paint transforms must be pre-concatenated onto the
current transform, matching the previous ttf-parser
Transform::combine(self.transform, transform) behaviour and skrifa's own
CollectFillGlyphPainter.

nicoburns committed:
Fix SVG-in-font lookup to use the record covering the glyph

The single-glyph vs multi-glyph document heuristic was checking the
first record in the SVG table instead of the record that actually
contains the requested glyph, picking the wrong branch for fonts whose
records differ (e.g. Noto Color Emoji).

nicoburns committed:
Avoid panics on malformed color fonts in COLR painting

Replace unwraps and direct slice indexing with graceful fallbacks:
- palette lookups fall back to the foreground color instead of
  panicking on a missing/malformed CPAL table or an out-of-range
  palette index
- clip glyph outlining returns early on a draw error, matching the
  old ttf-parser behaviour

nicoburns committed:
Respect CPAL palette record base offset

CPAL color records for all palettes live in a single shared array;
palette entry indices must be offset by the start index of the palette
from color_record_indices rather than indexing the array directly. This
is usually 0 for the first palette, but not guaranteed to be.

nicoburns committed:
Add pop_layer function

Signed-off-by: Nico Burns <nico@nicoburns.com>

nicoburns committed:
Support variable fonts

Make font metrics variation-aware

Resolve the requested font-variation-settings to a normalized location
in load_font so that skrifa applies MVAR deltas to ascent, descent,
x-height, underline and strikeout metrics, and so the manual
subscript/superscript (SBYO/SPYO) delta handling actually runs.
Metrics are cached per Font spec (which includes variations), so
caching remains correct.

Also drop the leftover commented-out ttf-parser code in outline():
with skrifa, the default (empty) location already resolves to the
default value of every axis.

Cache variable font outlines per variation set

Previously any span with explicit variations (or auto optical sizing on
a font with an opsz axis) bypassed the glyph outline cache entirely,
re-extracting the outline for every glyph occurrence.

Instead, compute the effective variation list (explicit settings plus
auto opsz) once per glyph via a shared helper, include it in the cache
key, and merge outline_with_variations into a single cached outline
function. An empty variation list resolves to the default location, so
non-variable fonts behave as before.

Apply font variations to COLR color glyphs

COLR glyphs were always painted at the default variation location,
ignoring font-variation-settings. Resolve the effective variations to a
normalized location and pass it to skrifa paint(), which resolves all
COLR ItemVariationStore deltas (paint transforms, gradient stops, clip
boxes). The GlyphPainter now also draws clip glyph outlines at the same
location so the base outlines vary too.

Color glyph trees are cached per variation set, like outlines.

nicoburns committed:
Hoist shaper construction out of the loop layout.rs

Signed-off-by: Nico Burns <nico@nicoburns.com>

nicoburns committed:
Optimize new snapshots with oxipng

Signed-off-by: Nico Burns <nico@nicoburns.com>

akx commented on crates/usvg/src/text/flatten.rs:
```diff
@@ -200,11 +207,12 @@ pub(crate) fn flatten(text: &mut Text, cache: &mut Cache) -> Option<(Group, NonZ
 }
 
-impl ttf_parser::OutlineBuilder for PathBuilder {
+impl OutlinePen for PathBuilder {
```
FWIW, via #1097, this and colr.rs's Builder are practically the same thing.

nicoburns committed:
Fix unbalanced clip groups in COLR painting

Skrifa pairs push_clip_glyph with pop_clip (including via the default
fill_glyph decomposition), but push_clip_glyph never opened an XML
element while pop_clip unconditionally closed one. Each painted layer
therefore closed an ancestor element, truncating the generated SVG and
making Tree::from_data fail for color glyphs with more than a few
layers.

This went unnoticed because the Noto COLR test font also contains an
SVG table, so rendering silently fell back to the SVG glyph path.

push_clip_glyph now always opens a clip group with the glyph outline
(an empty path when outlining fails), keeping pop_clip balanced.

The colrv0/colrv1 snapshots change because COLR rendering now takes
priority over the SVG table again; the output matches main.

nicoburns committed:
Use skrifa ColorPalettes API for CPAL lookups

ColorPalettes handles per-palette record offsets internally, so the
manual color_record_indices base-offset handling is no longer needed.

nicoburns committed:
Use axes().get_by_tag() for opsz axis checks

Replaces the manual fvar table iteration in has_opsz_axis and the
shaping auto-opsz check, and drops the harfrust FontRef usage from
flatten.rs (skrifa and harfrust share the same underlying types).

nicoburns committed:
Skip building a ShaperInstance when there are no variations

The instance only exists to apply variation settings, so avoid the
setup cost for non-variable text (the common case).

nicoburns committed:
Reject fonts with an out-of-range unitsPerEm

ttf-parser enforced 16..=16384 when parsing the head table; skrifa
does not, so add the check to load_font to avoid nonsensical metrics
from malformed fonts.

nicoburns committed:
Round font metrics instead of truncating

skrifa reports metrics as f32; round when converting to the integer
font units used by ResolvedFont rather than truncating toward zero,
matching the integer values ttf-parser used to return. Also ignore
non-positive x-height values, like ttf-parser did.

nicoburns committed:
Implement the COLR clip stack model correctly

The painter previously reused the last outlined glyph (path_buf) as the
fill geometry, which only works when exactly one clip glyph is active.
The COLR imaging model instead has a proper clip stack: a fill paints
the intersection of all currently pushed clips.

- fill() now paints a rectangle covering the intersection of the
  tracked clip bounds (in root space), shaped by the enclosing SVG clip
  groups, instead of whatever happened to be in path_buf.
- push_clip_glyph/push_clip_box track their bounds on a clip stack.
- fill_glyph() is implemented directly (fill the glyph outline with the
  brush), avoiding the default clip-then-fill decomposition. This also
  restores the exact rendering of the old ttf-parser painter for
  COLRv0, so colrv0.png reverts to the version on main.

The colrv1 snapshot changes slightly at glyph edges: leaf fills no
longer pass through a clip of the same shape, so edge anti-aliasing is
no longer applied twice.

akx commented on crates/usvg/src/text/colr.rs:
```diff
@@ -130,17 +147,15 @@ impl<'a> GlyphPainter<'a> {
         // we will see the actual spreadMode. We need to account for that somehow.
```
I think this comment isn't relevant anymore since the new data model doesn't have x2/y2?

akx commented on crates/usvg/src/text/colr.rs:
```diff
@@ -245,30 +248,135 @@ impl GlyphPainter<'_> {
     }
+
+    fn palette_index_to_color(&self, palette_index: u16, alpha: f32) -> Color {
+        let lookup = || -> Option<Color> {
+            // We always use the first palette. `ColorPalettes` handles
+            // per-palette record offsets internally.
+            let palettes = self.font.color_palettes();
+            let palette = palettes.get(0)?;
+            let color = palette.colors().get(palette_index as usize)?;
+            Some(Color {
+                red: color.red,
+                blue: color.blue,
+                green: color.green,
+                alpha: color.alpha,
+            })
+        };
```
Why the lambda for the lookup?

akx commented on crates/usvg/src/text/flatten.rs:
```diff
@@ -21,6 +30,26 @@ fn resolve_rendering_mode(text: &Text) -> ShapeRendering {
+    span: &layout::Span,
+    glyph: &layout::PositionedGlyph,
+) -> Vec<FontVariation> {
+    let mut variations = span.variations.clone();
```
Are we worried about the clone for each glyph?

nicoburns replied:
I have created an issue for this.

akx commented on crates/usvg/src/text/layout.rs:
```diff
@@ -82,7 +85,7 @@ impl PositionedGlyph {
             // Right now, the top-left corner of the image would be placed in
             // on the "text cursor", but we want the bottom-left corner to be there,
             // so we need to shift it up and also apply the x/y offset.
```
Is this comment accurate given the change below?

akx commented on crates/usvg/src/text/layout.rs:
```diff
@@ -1408,13 +1435,16 @@ fn shape_text_with_font(
     fontdb.with_face_data(font.id, |font_data, face_index| -> Option<Vec<Glyph>> {
-        let mut rb_font = rustybuzz::Face::from_slice(font_data, face_index)?;
+        use harfrust::{Feature, ShaperData, ShaperInstance, Tag, UnicodeBuffer, Variation};
+        const OPSZ: Tag = Tag::from_be_bytes(*b"opsz");
```
I think I saw this in `flatten.rs` too...

akx commented on crates/usvg/src/text/mod.rs:
```diff
@@ -10,11 +10,22 @@ use self::layout::DatabaseExt;
 
+/// The ID of a glyph within a font
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct GlyphId(pub u16);
```
Is u16 really enough, if skrifa does u32?

behdad replied:
u32 please. 24-bit glyph-ids are coming to the ISO OpenFontFormat version 5 that will be released to public any day now.

https://www.iso.org/obp/ui/#iso:std:iso-iec:14496:-22:ed-5:v1:en

nicoburns replied:
Done!

akx reviewed (commented):
Some minor comments from a human-eyes review, not sure if they're relevant or actionable :)

nicoburns committed:
Use u32 for GlyphId

LaurenzV commented on crates/usvg/Cargo.toml:
```diff
@@ -36,11 +36,9 @@ simplecss = "0.2"
-# Note: ttf-parser is re-exported from rustybuzz, but we need gvar-alloc for variable fonts
-# with many variation axes (like Roboto Flex which has 13 axes)
-ttf-parser = { version = "0.25.1", features = ["gvar-alloc"], optional = true }
+fontdb = { git = "https://github.com/nicoburns/fontdb", rev = "2b04f122d2bd76177c16ea5c1c05dd85da7ac18a", default-features = false, optional = true }
```
Let's maybe make the fontdb release before merging.

LaurenzV commented on crates/usvg/src/text/layout.rs:
```diff
@@ -1201,51 +1204,67 @@ fn form_glyph_clusters(glyphs: &[Glyph], text: &str, font_size: f32) -> GlyphClu
+            let metrics = font.metrics(Size::unscaled(), &location);
 
-            let units_per_em = NonZeroU16::new(font.units_per_em())?;
+            // Reject fonts with an out-of-range unitsPerEm, like `ttf-parser` did.
+            if !(16..=16384).contains(&metrics.units_per_em) {
+                return None;
+            }
```
I don't think this kind of edge case should be handled in resvg. It's fine if the behavior is different for those.

akx replied:
IIRC, the issue is there's math below that assumes this to be true.

LaurenzV commented on crates/usvg/src/text/flatten.rs:
```diff
@@ -4,15 +4,24 @@
 use crate::text::colr::GlyphPainter;
 use crate::*;
 
+const OPSZ: Tag = Tag::from_be_bytes(*b"opsz");
```
I think that const exists twice now, maybe reuse?

LaurenzV commented on crates/usvg/src/text/flatten.rs:
```diff
@@ -395,9 +384,17 @@ impl DatabaseExt for Database {
+            let font = skrifa::FontRef::from_index(data, face_index).ok()?;
+
+            // COLR paints (transforms, gradient stops, clip boxes) and the
+            // referenced glyph outlines can all vary in a variable font.
```
```suggestion
```

Don't think we need to explicitly comment this.

LaurenzV commented on crates/usvg/src/parser/converter.rs:
```diff
@@ -204,11 +204,45 @@ impl Cache {
+        glyph: GlyphId,
+        variations: &[FontVariation],
+    ) -> Option<tiny_skia_path::Path> {
+        let key = (font, glyph, variations.to_vec());
```
Maybe let's add a TODO to use smallvecs for the variations.

LaurenzV commented on crates/usvg/src/text/colr.rs:
```diff
@@ -184,50 +205,61 @@ impl<'a> GlyphPainter<'a> {
+        _color_stops: &[ColorStop],
+        _extend: Extend,
+    ) {
         println!("Warning: sweep gradients are not supported.");
```
Let's still remove this even though it already existed before.

LaurenzV reviewed (approved):
Thanks for taking this on!

nicoburns committed:
Remove COLR comment

Co-authored-by: Laurenz Stampfl <47084093+LaurenzV@users.noreply.github.com>

nicoburns committed:
Deduplicate OPSZ constant

nicoburns committed:
Use fontdb 0.24

Signed-off-by: Nico Burns <nico@nicoburns.com>

nicoburns committed:
Remove println's from colr module

nicoburns committed:
Update TODO comments

Signed-off-by: Nico Burns <nico@nicoburns.com>
